File: ios/MappySceneViewManager.m
//
//  MappySceneViewManager.m
//  SnowMappyRN
//
//  Created by Afsar Shakeer on 2023-08-07.
//

#import <Foundation/Foundation.h>
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(MappySceneViewManager, RCTViewManager)

RCT_EXTERN_METHOD(
  setScene:(nonnull NSNumber *)node
  venueInfo:(nonnull NSDictionary *)info
)

RCT_EXTERN_METHOD(
  rotateToNorth:(nonnull NSNumber *)node
)

RCT_EXTERN_METHOD(
  setInitialViewpoint:(nonnull NSNumber *)node
)

@end

File: ios/MappySceneViewManager.swift
//
//  MappySceneViewManager.swift
//  SnowMappyRN
//
//  Created by Afsar Shakeer on 2023-08-07.
//

import Foundation

@objc(MappySceneViewManager)
final class MappySceneViewManager: RCTViewManager {
  override func view() -> UIView! {
    return MappySceneViewContainer()
  }
  
  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  @objc
  func setScene(_ node: NSNumber, venueInfo: NSDictionary) {
      DispatchQueue.main.async {
        self.mappySceneViewContainer(node).setScene(venueInfo)
      }
  }
  
  @objc
  func rotateToNorth(_ node: NSNumber) {
    DispatchQueue.main.async {
      self.mappySceneViewContainer(node).rotateToNorth()
    }
  }
  
  @objc
  func setInitialViewpoint(_ node: NSNumber) {
    DispatchQueue.main.async {
      self.mappySceneViewContainer(node).setInitialViewpoint()
    }
  }
  
  private func mappySceneViewContainer(_ node: NSNumber) -> MappySceneViewContainer {
    self.bridge.uiManager.view(forReactTag: node) as! MappySceneViewContainer
  }
}
